### PR TITLE
ci: [IOPLT-000] Update `Jest` running condition

### PR DESCRIPTION
**Pull request by LeleDallas**
## Short description
This pull request updates the test execution logic in the GitHub Actions workflow to optimize test runs for pull requests. Now, when a pull request is opened, only tests affected by the changes are run, improving CI efficiency.

## List of changes proposed in this pull request
- Modified the `run-test` step in `.github/workflows/staticcheck.yaml` to run only tests affected by changes (`yarn test:ci --changedSince=origin/${{ github.base_ref }}`) for pull request events, while keeping full sharded test runs for other events.

## How to test
- Apply some changes into this PR verify that the CI runs only the tests related to the modified files using `--changedSince`, without `sharding`, so it completes faster. 
- When merging to main or testing via `merge_group`, confirm that the full test suite runs with `sharding` and that coverage is still uploaded correctly to Codecov.

The expected behavior is fast feedback on PRs while maintaining full coverage checks on main.

### CODE REVIEW

**Comment by dmnplb** [comment]
There's a real risk of encountering an annoying edge case. For example, when we update the DS version, we often need to update lots of snapshots to ensure that all the tests pass. With the current setup, we only realise this when we add it to the merge queue.

What if we included the edits to `package.json` or `yarn.lock` as a condition to trigger a full test suite?

**Comment by LeleDallas** [comment]
> There's a real risk of encountering an annoying edge case. For example, when we update the DS version, we often need to update lots of snapshots to ensure that all the tests pass. With the current setup, we only realise this when we add it to the merge queue.
> 
> What if we included the edits to `package.json` or `yarn.lock` as a condition to trigger a full test suite?

Makes sense to me! Addressed here 63481d6d6dff1ca31d1bd5c2b7085d4efda140f5

**Review by dmnplb — APPROVED** [review]
To me, it seems like a really good DX improvement. I haven't really tested it, but you've fixed the edge case I was thinking about, so LGTM! ✌🏻

File: .github/workflows/static-checks.yaml [hunk 1]
```diff
@@ -4,6 +4,32 @@ on:
   - merge_group
   - workflow_call
 jobs:
+  configure:
+    runs-on: ubuntu-latest
```
Comment (dmnplb) [inline .github/workflows/static-checks.yaml:11]: What about adding a more descriptive name to the job?
```suggestion
    name: Compute shard matrix
    runs-on: ubuntu-latest
```

File: .github/workflows/static-checks.yaml [hunk 2]
```diff
@@ -13,6 +39,8 @@ jobs:
     steps:
       - id: checkout
         uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
+        with:
+          fetch-depth: 0
```
Comment (dmnplb) [inline .github/workflows/static-checks.yaml]: Do we really need the full `git` history in this step?
```suggestion
```

File: .github/workflows/static-checks.yaml [hunk 3]
```diff
@@ -52,7 +82,35 @@ jobs:
           name: api-client
           path: definitions/
       - id: run-test
-        run: yarn test:ci --shard=${{ matrix.shard }}/${{ strategy.job-total }}
+        shell: bash
+        run: |
+          if [ "${{ github.event_name }}" = "pull_request" ]; then
+            echo "Checking if lockfile or critical files changed..."
+
+            CHANGED_FILES=$(git diff --name-only origin/${{ github.base_ref }}...HEAD)
```
Comment (dmnplb) [inline .github/workflows/static-checks.yaml]: What about filtering the changed files to avoid starting testing process on files that we're not interested about (`README`, `*.yml`, etc…). I'd also include `*.json` because the locale files could disrupt tests.
```suggestion
            CHANGED_FILES=$(git diff --name-only origin/${{ github.base_ref }}...HEAD \
    | grep -E '\.(ts|tsx|js|jsx)$' | wc -l)
```

Comment (LeleDallas) [reply .github/workflows/static-checks.yaml]: Makes sense to me! But your suggestion is wrong since it doesn't include `json` files

File: .github/workflows/static-checks.yaml (hunk 1 above)
Comment (LeleDallas) [reply .github/workflows/static-checks.yaml:11]: Addressed here bd72122e1dd100c819f15f2aeaf98a6260cf810e

File: .github/workflows/static-checks.yaml (hunk 2 above)
Comment (LeleDallas) [reply .github/workflows/static-checks.yaml]: Addressed here 424969f23c040ee80a549f5d97e0cd07eb30b102

File: .github/workflows/static-checks.yaml (hunk 3 above)
Comment (LeleDallas) [reply .github/workflows/static-checks.yaml]: I made it more readable using git pattern here 99ce51057211e276e94faacfe60006f27c57fefb

**Review by CrisTofani — APPROVED** [review]
LGTM

**Review by CrisTofani — CHANGES_REQUESTED** [review]
Checking the Github config we cannot define dynamically the required jobs to pass before merging a PR. So I would suggest at this point to avoid the shard since in most common cases we'd have just few files to test and so the execution would summarize into one single shard execution. What do you think?

**Comment by LeleDallas** [comment]
> Checking the Github config we cannot define dynamically the required jobs to pass before merging a PR. So I would suggest at this point to avoid the shard since in most common cases we'd have just few files to test and so the execution would summarize into one single shard execution. What do you think?

Update running condition here c9fc872fdc53b2a410a509d90ee2347d0b7064cc

**Review by CrisTofani — DISMISSED** [review]
LGTM

**Review by CrisTofani — APPROVED** [review]
Checked latest commits LGTM
